dask/highlevelgraph.py

```diff
@@ -119,27 +119,43 @@ def from_collections(cls, name, layer, dependencies=()):
         ...     graph = HighLevelGraph.from_collections(name, layer, dependencies=[self])
         ...     return new_collection(name, graph)
         """
-        layers = {name: layer}
-        deps = {}
-        deps[name] = set()
-        for collection in toolz.unique(dependencies, key=id):
-            if is_dask_collection(collection):
-                graph = collection.__dask_graph__()
-                if isinstance(graph, HighLevelGraph):
-                    layers.update(graph.layers)
-                    deps.update(graph.dependencies)
-                    with ignoring(AttributeError):
-                        deps[name] |= set(collection.__dask_layers__())
-                else:
-                    try:
-                        [key] = collection.__dask_layers__()
-                    except AttributeError:
-                        key = id(graph)
-                    layers[key] = graph
-                    deps[name].add(key)
-                    deps[key] = set()
-            else:
-                raise TypeError(type(collection))
+        if any(toolz.remove(is_dask_collection, toolz.unique(dependencies, key=id))):
+            raise TypeError
+
+        graphs_and_collections = [
+            (collection.__dask_graph__(), collection)
+            for collection in toolz.unique(dependencies, key=id)
+        ]
+
+        hlgs = filter(
+            lambda x: isinstance(x[0], HighLevelGraph), graphs_and_collections
+        )
+        others = filter(
+            lambda x: not isinstance(x[0], HighLevelGraph), graphs_and_collections
+        )
+
+        layers = toolz.merge((graph.layers for graph, _ in hlgs))
+        deps = toolz.merge((graph.dependencies for graph, _ in hlgs))
```

Comment:
Since you're iterating over `hlgs` multiple times, you're duplicating the filtering done above (since `filter` is an iterator, not a list). Would be better to filter once as a list than multiple times.

Reply:
Turns out since it is an iterator, I was only getting any values on the first pass :facepalm: